src/Equations/SPIHandler.cpp

```diff
@@ -87,44 +109,74 @@ SPIHandler::SPIHandler(FVM::Grid *g, FVM::UnknownQuantityHandler *u, len_t *Z, l
     // Get number of grid points and number of shards
     this->nr=rGrid->GetNr();
     this->nShard=unknowns->GetUnknown(id_Yp)->NumberOfMultiples();
+    
+    this->NZ=NZ;
 
     // Memory allocation
     AllocateQuantities();
     
+    // Ablation cloud quantities
+    this->TDrift=TDrift;
+    this->T0Drift=T0Drift;
+    this->DeltaYDrift=DeltaYDrift;
+    this->RmDrift=RmDrift;
+    this->ZavgDriftArray=ZavgDriftArray;
+    
+    
     // Initialize rCoordPrevious to the radial coordinate at the plasma edge 
     // to use as a starting guess if rCoord must be solved for numerically
     for(len_t ip=0;ip<nShard;ip++)
         rCoordPPrevious[ip]=rGrid->GetR_f(nr-1);
 
-    // Calculate pellet molar mass, molar volume and density
+    // Calculate pellet molar mass, molar volume, density and average charge of the ablation cloud
+    // The later is calculated based on a user-given lookup table (ZavgArray) for the average charge
+    // of the relevant species (specified in Zs and isotopesDrift), as the ADAS rates are not valid 
+    // for the conditions inside the ablation cloud
     real_t molarMass=0;
     real_t solidDensity=0;
-    real_t *pelletDeuteriumFraction=new real_t[nShard];
+    real_t ZavgDriftList=0;
     for(len_t ip=0;ip<nShard;ip++){
         pelletMolarMass[ip]=0;
         pelletMolarVolume[ip]=0;
+        ZavgDrift[ip]=0;
     }
     
     len_t offset=0;
+    len_t counter1=0;
+    len_t counter2=0;
     for(len_t iZ=0;iZ<NZ;iZ++){
         if(molarFraction[offset]>=0){
             for(len_t i=0;i<nMolarMassList;i++){
                 if(Z[iZ]==ZMolarMassList[i] && isotopes[iZ]==isotopesMolarMassList[i]){
                     molarMass=molarMassList[i];
                 }
+                else{counter1++;}
             }
+            if(counter1==nMolarMassList){throw DREAMException("SPIHandler: Pellet type is not recognized. Currently only neon and deuterium pellets are supported. To support other types fill in the material data in src/Equations/SPIHandler.cpp and py/DREAM/Settings/Equations/SPI.py.");}
+            counter1=0;
+            
             for(len_t i=0;i<nSolidDensityList;i++){
                 if(Z[iZ]==ZSolidDensityList[i] && isotopes[iZ]==isotopesSolidDensityList[i]){
                     solidDensity=solidDensityList[i];
                 }
             }
+            for(len_t i=0;i<nZavgDrift;i++){
+                if(Z[iZ]==ZsDrift[i] && isotopes[iZ]==isotopesDrift[i]){
+                    ZavgDriftList=ZavgDriftArray[i];
+                }
+            }
             for(len_t ip=0;ip<nShard;ip++){
 		        pelletMolarMass[ip]+=molarMass*molarFraction[offset+ip];
 		        pelletMolarVolume[ip]+=molarMass/solidDensity*molarFraction[offset+ip];
-		 
-		        if(Z[iZ]==1 && isotopes[iZ]==2){
+                
+		        if(Z[iZ]==1 && isotopes[iZ]==2)
 		            pelletDeuteriumFraction[ip]+=molarFraction[offset+ip];
```

Comment:
Why is this ``+=`` and not just ``=``? At this point, ``pelletDeuteriumFraction`` is uninitialized and could contain garbage values. It must therefore either be properly initialized first, or we should turn ``+=`` into ``=``.